### PR TITLE
Migration all the functions  from VaadinIcon to IconFactory

### LINKED ISSUE #69: Proposal: Migrate all routerLink functions
In order to support both VaadinIcon and custom Icon collections I suggest to migrate all functions like
`fun (@VaadinDsl HasComponents).routerLink(icon: VaadinIcon?, ...`
from VaadinIcon to IconFactory
`fun (@VaadinDsl HasComponents).routerLink(icon: IconFactory?, ...`

Please review a new code and example of usage in [Demo project for Vaadin routerLink](https://github.com/vlipovetskii/vaadin-router-link)

Please note: A new code is backward-compatible.

@mvysny if it's ok for you, I will prepare a PR.

### PR DESCRIPTION
**Pull request by vlipovetskii**
Fixes #69

### CODE REVIEW

File: karibu-dsl/src/main/kotlin/com/github/mvysny/karibudsl/v10/Router.kt [hunk 1]
```diff
@@ -2,39 +2,39 @@ package com.github.mvysny.karibudsl.v10
...
 import kotlin.reflect.KClass
 import com.vaadin.flow.router.RouteParameters
 
 @VaadinDsl
-public fun (@VaadinDsl HasComponents).routerLink(icon: VaadinIcon? = null, text: String? = null, viewType: KClass<out Component>,
+public fun (@VaadinDsl HasComponents).routerLink(icon: IconFactory? = null, text: String? = null, viewType: KClass<out Component>,
                                                  block: (@VaadinDsl RouterLink).() -> Unit = {}): RouterLink {
     val link = RouterLink(null as String?, viewType.java)
-    if (icon != null) link.icon(icon)
+    if (icon != null) link.init(icon.create())
```
Comment (mvysny) [inline karibu-dsl/src/main/kotlin/com/github/mvysny/karibudsl/v10/Router.kt]: Hi, I've updated the `icon{}` DSL to also accept `IconFactory`. It should work as-is, could you please update & revert this line?

File: karibu-dsl/src/main/kotlin/com/github/mvysny/karibudsl/v10/Router.kt [hunk 2]
```diff
@@ -2,39 +2,39 @@ package com.github.mvysny.karibudsl.v10
...
 }
 
 @JvmName("routerLinkWithParams")
 @VaadinDsl
-public fun (@VaadinDsl HasComponents).routerLink(icon: VaadinIcon? = null, text: String? = null, viewType: KClass<out Component>,
+public fun (@VaadinDsl HasComponents).routerLink(icon: IconFactory? = null, text: String? = null, viewType: KClass<out Component>,
                                                         parameters: Map<String, String>, block: (@VaadinDsl RouterLink).() -> Unit = {}): RouterLink {
     val link = RouterLink(null as String?, viewType.java, RouteParameters(parameters))
-    if (icon != null) link.icon(icon)
+    if (icon != null) link.init(icon.create())
```
Comment (mvysny) [inline karibu-dsl/src/main/kotlin/com/github/mvysny/karibudsl/v10/Router.kt]: Same here, and also in 2 other places.

**Review by mvysny — CHANGES_REQUESTED** [review]

**Comment by mvysny** [comment]
Fixes #69

File: karibu-dsl/src/main/kotlin/com/github/mvysny/karibudsl/v10/Router.kt (hunk 1 above)
Comment (vlipovetskii) [reply karibu-dsl/src/main/kotlin/com/github/mvysny/karibudsl/v10/Router.kt]: Hi! I did it.

**Review by mvysny — APPROVED** [review]

**Comment by mvysny** [comment]
Perfect, awesome work, thank you

**Pull request merged**